File: ActionUI/Views/ContentUnavailableView.swift
/*
 Sample JSON for ContentUnavailableView:
 {
   "type": "ContentUnavailableView",
   "id": 1,              // Optional: Non-zero positive integer for runtime programmatic interaction
   "properties": {
     "title": "No Results",              // Required: String for the primary message
     "systemImage": "magnifyingglass",   // Optional: String for SF Symbol, defaults to nil (no image)
     "description": "Try a different search term." // Optional: String for secondary descriptive text, defaults to nil
   }
 }

 Search variant (shows "No results for <query>" with magnifying glass icon):
 {
   "type": "ContentUnavailableView",
   "id": 2,
   "properties": {
     "search": true,      // Optional: Boolean. When true, uses the built-in search appearance.
                           //   The search query text can be set via setElementValue to display "No results for <query>".
                           //   When search is true, title/systemImage/description are ignored.
     "query": "planets"   // Optional: String for the search query shown in the message. Can also be set at runtime via value.
   }
 }

 // Note: ContentUnavailableView is the standard SwiftUI view for displaying empty states, missing data,
 // or no-results conditions. Available on macOS 14.0+, iOS 17.0+.
 // Baseline View properties (padding, hidden, foregroundColor, font, background, frame, opacity, cornerRadius, actionID, disabled)
 // and additional View protocol modifiers are inherited and applied via
 // ActionUIRegistry.shared.applyViewModifiers(to: baseView, properties: element.properties).
*/

import SwiftUI

struct ContentUnavailableView: ActionUIViewConstruction {
    static var valueType: Any.Type { String.self }

    static var validateProperties: ([String: Any], any ActionUILogger) -> [String: Any] = { properties, logger in
        var validatedProperties = properties

        // Validate title
        if properties["title"] != nil && !(properties["title"] is String) {
            logger.log("ContentUnavailableView title must be a String; ignoring", .warning)
            validatedProperties["title"] = nil
        }

        // Validate systemImage
        if properties["systemImage"] != nil && !(properties["systemImage"] is String) {
            logger.log("ContentUnavailableView systemImage must be a String; ignoring", .warning)
            validatedProperties["systemImage"] = nil
        }

        // Validate description
        if properties["description"] != nil && !(properties["description"] is String) {
            logger.log("ContentUnavailableView description must be a String; ignoring", .warning)
            validatedProperties["description"] = nil
        }

        // Validate search
        if properties["search"] != nil && !(properties["search"] is Bool) {
            logger.log("ContentUnavailableView search must be a Boolean; ignoring", .warning)
            validatedProperties["search"] = nil
        }

        // Validate query
        if properties["query"] != nil && !(properties["query"] is String) {
            logger.log("ContentUnavailableView query must be a String; ignoring", .warning)
            validatedProperties["query"] = nil
        }

        return validatedProperties
    }

    static var buildView: (any ActionUIElementBase, ViewModel, String, [String: Any], any ActionUILogger) -> any SwiftUI.View = { element, model, windowUUID, properties, logger in
        let isSearch = properties["search"] as? Bool ?? false

        if isSearch {
            let query = model.value as? String ?? properties["query"] as? String ?? ""
            if query.isEmpty {
                return SwiftUI.ContentUnavailableView.search
            } else {
                return SwiftUI.ContentUnavailableView.search(text: query)
            }
        }

        let title = model.value as? String ?? properties["title"] as? String ?? "No Content"
        let description = properties["description"] as? String
        let systemImage = properties["systemImage"] as? String

        if let systemImage {
            if let description {
                return SwiftUI.ContentUnavailableView(title, systemImage: systemImage, description: SwiftUI.Text(description))
            } else {
                return SwiftUI.ContentUnavailableView(title, systemImage: systemImage)
            }
        } else {
            if let description {
                return SwiftUI.ContentUnavailableView {
                    SwiftUI.Label(title, systemImage: "exclamationmark.triangle")
                        .labelStyle(.titleOnly)
                } description: {
                    SwiftUI.Text(description)
                }
            } else {
                return SwiftUI.ContentUnavailableView {
                    SwiftUI.Label(title, systemImage: "exclamationmark.triangle")
                        .labelStyle(.titleOnly)
                }
            }
        }
    }

    static var initialValue: (ViewModel) -> Any? = { model in
        if let storedValue = model.value as? String {
            return storedValue
        }
        // For search variant, use query; otherwise use title
        if model.validatedProperties["search"] as? Bool == true {
            return model.validatedProperties["query"] as? String ?? ""
        }
        return model.validatedProperties["title"] as? String ?? "No Content"
    }
}

File: ActionUIWeb/src/Views/ContentUnavailableView.js
// ContentUnavailableView.js — ContentUnavailableView element.
// Web analog of ActionUI/Views/ContentUnavailableView.swift and ActionUIAndroid
// Views/ContentUnavailableView.kt.
//
// The standard empty-state view: a centered hero glyph over a bold title and an
// optional secondary description - SwiftUI's ContentUnavailableView. The web
// renders real icons (the shared SymbolIcon glyph seam), where Android downgrades
// the icon to text on its asset path; the web already has the glyph path.
//
// Properties (ContentUnavailableView.swift):
//   title       String (the primary message; default "No Content")
//   systemImage String SF Symbol (default "exclamationmark.triangle" when absent,
//               mirroring Swift's fallback Label icon)
//   description String (secondary text; omitted when absent)
//   search      Bool - the built-in search appearance (magnifyingglass +
//               "No Results"); when true, title/systemImage/description are ignored
//   query       String - the search term shown in the message
//
// Value bridge (Apple's, valueType "string"): the runtime value is the title
// (non-search) or the query (search). get/setElementValue round-trips it; a host
// setString("planets") on a search view re-renders it as "No Results for ...".

import { register } from "../Common/ActionUIRegistry.js";
import { systemSymbolGlyph } from "../Helpers/SymbolIcon.js";

register("ContentUnavailableView", {
    valueType: "string",

    // Mirrors ContentUnavailableView.swift validateProperties (verbatim warnings).
    validateProperties: (properties, logger) => {
        const validated = { ...properties };
        if (validated.title !== undefined && typeof validated.title !== "string") {
            logger.log("ContentUnavailableView title must be a String; ignoring", "warning");
            delete validated.title;
        }
        if (validated.systemImage !== undefined && typeof validated.systemImage !== "string") {
            logger.log("ContentUnavailableView systemImage must be a String; ignoring", "warning");
            delete validated.systemImage;
        }
        if (validated.description !== undefined && typeof validated.description !== "string") {
            logger.log("ContentUnavailableView description must be a String; ignoring", "warning");
            delete validated.description;
        }
        if (validated.search !== undefined && typeof validated.search !== "boolean") {
            logger.log("ContentUnavailableView search must be a Boolean; ignoring", "warning");
            delete validated.search;
        }
        if (validated.query !== undefined && typeof validated.query !== "string") {
            logger.log("ContentUnavailableView query must be a String; ignoring", "warning");
            delete validated.query;
        }
        return validated;
    },

    // The runtime value: the query in search mode, otherwise the title.
    initialValue: (element, properties) => {
        if (properties.search === true) return properties.query ?? "";
        return properties.title ?? "No Content";
    },

    buildView: (element, properties, ctx) => {
        const isSearch = properties.search === true;
        const node = document.createElement("div");
        node.className = "aui-content-unavailable";

        const render = (value) => {
            node.replaceChildren();
            const glyphName = isSearch ? "magnifyingglass"
                : (typeof properties.systemImage === "string" ? properties.systemImage : "exclamationmark.triangle");
            const glyph = systemSymbolGlyph(glyphName, { imageScale: "large" }, ctx.logger, (n) =>
                `ContentUnavailableView systemImage '${n}' has no SF->Material mapping; icon omitted.`);
            glyph.classList.add("aui-content-unavailable-icon");
            node.appendChild(glyph);

            const title = document.createElement("span");
            title.className = "aui-content-unavailable-title";
            const description = document.createElement("span");
            description.className = "aui-content-unavailable-description";

            if (isSearch) {
                const query = String(value ?? "");
                title.textContent = "No Results";
                description.textContent = query.length > 0
                    ? `No results for "${query}". Check the spelling or try a new search.`
                    : "Check the spelling or try a new search.";
            } else {
                title.textContent = String(value ?? "") || "No Content";
                if (typeof properties.description === "string") {
                    description.textContent = properties.description;
                }
            }

            node.appendChild(title);
            if (description.textContent) node.appendChild(description);
        };

        const initial = isSearch ? (properties.query ?? "") : (properties.title ?? "No Content");
        let current = initial;
        render(current);

        if (element.id > 0) {
            ctx.model.bind(element.id, {
                getValue: () => current,
                setValue: (value) => { current = String(value ?? ""); render(current); },
            });
        }
        return node;
    },
});
